File: PAT_Basic_Level_by_ruby/1097 矩阵行平移/1097 矩阵行平移.rb
# 1097 矩阵行平移
#
# 题目描述：
#   给定一个 n\times n 的整数矩阵。对任一给定的正整数 k<n，我们将矩阵的奇数行的元素整体向右依次平移 1、……、k、1、……、k、…… 个位置，平移空出的位置用整数 x 补。你需要计算出结果矩阵的每一列元素的和。
#
# 实现原理：
#   对矩阵的奇数行进行循环右移；不同行移动不同的位数；计算各列的和并找出最大值
#

if __FILE__ == $PROGRAM_NAME
  n, k, x = gets.split.map(&:to_i)
  matrix = n.times.map { gets.split.map(&:to_i) }
  shift = 1

  (0...n).step(2) do |i|
    original = matrix[i].dup
    matrix[i] = Array.new(n, x)
    (shift...n).each { |j| matrix[i][j] = original[j - shift] }
    shift += 1
    shift = 1 if shift > k
  end

  puts (0...n).map { |column| matrix.sum { |row| row[column] } }.join(' ')
end

File: PAT_Basic_Level_by_c/1097 矩阵行平移/1097 矩阵行平移.c
/*
 * 题目：1097 矩阵行平移
 * 实现原理：
 *   对矩阵的奇数行（0行、2行、4行...）进行向右平移操作。
 *   第1次平移1位，第2次平移2位，...，第k次平移k位，然后重复。
 *   平移后的空位填充数字x。
 *   最后计算每列的和并输出。
 *   算法步骤：
 *   1. 读取n*n矩阵
 *   2. 对奇数行进行平移：将行内容保存到temp，前shift位填x，后面填原内容
 *   3. 计算每列的和并输出
 */
#include <stdio.h>

int main() {
    int n, k, x;        // n:矩阵阶数; k:平移位数循环周期; x:空位填充值
    scanf("%d %d %d", &n, &k, &x);
    int matrix[100][100];
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            scanf("%d", &matrix[i][j]);
        }
    }
    int shift = 1;
    for (int i = 0; i < n; i += 2) {    // 只处理奇数行（第1、3、5...行）
        int temp[100];
        for (int j = 0; j < n; j++) temp[j] = matrix[i][j];     // 暂存整行
        for (int j = 0; j < n; j++) {
            if (j < shift) matrix[i][j] = x;        // 前shift位填入x
            else matrix[i][j] = temp[j - shift];    // 其余位右移
        }
        shift++;
        if (shift > k) shift = 1;       // 平移位数在1~k之间循环
    }
    for (int j = 0; j < n; j++) {       // 逐列求和
        int sum = 0;
        for (int i = 0; i < n; i++) sum += matrix[i][j];
        if (j > 0) printf(" ");
        printf("%d", sum);
    }
    printf("\n");
    return 0;
}